content/blog/2019-11-dtplyr-1-0-0/index.Rmarkdown

```diff
@@ -0,0 +1,228 @@
+---
+title: dtplyr 1.0.0
+author: Hadley Wickham
+date: '2019-11-08'
+slug: dtplyr-1-0-0
+description: |
+  A total rewrite of dtplyr is now available on CRAN; it now performs
+  computation lazily (like dbplyr), making it much more performant.
+categories:
+  - package
+tags:
+  - dtplyr
+  - dplyr
+  - tidverse
+photo:
+  url: https://unsplash.com/photos/h13Y8vyIXNU
+  author: Jay Ruzesky
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+I'm very excited to announce that [dtplyr](https://dtplyr.tidyverse.org) 1.0.0 is now on CRAN. dtplyr provides a [data.table](http://r-datatable.com/) backend for dplyr, allowing you to write dplyr code that is automatically translated to the equivalent, but usually much faster, data.table code.  
+
+This version of dtplyr is a complete rewrite which allows dtplyr to generate significantly more performant translations. dtplyr now works like [dbplyr](https://dbplyr.tidyverse.org), where computation only happens when requested by  `as.data.table()`, `as.data.frame()` or `as_tibble()` (this idea can also be seen in the [table.express](https://github.com/asardaes/table.express) and [rqdatatable](https://github.com/WinVector/rqdatatable/) packages). Unfortunately, this rewrite breaks pretty much all existing uses of dtplyr. But frankly, the prior version of dtplyr was not very good and few people used it, so a major overhaul should break little code in the wild.
+
+In this article, I'll introduce you to the basic usage of dtplyr, talk about some of the performance implications, and show off some of the translations that I'm most proud of.
+
+## Usage
+
+To use dtplyr, you must at least load dtplyr and dplyr. You might also want to load [data.table](http://r-datatable.com/) so you can access the other goodies that it provides:
+
+```{r}
+library(data.table)
+library(dtplyr)
+library(dplyr, warn.conflicts = FALSE)
+```
+
+Then use `lazy_dt()` to create a "lazy" data table that tracks the operations performed on it:
+
+```{r}
+mtcars2 <- lazy_dt(mtcars)
+```
+
+You can use dplyr verbs with this object as if it was a data frame. But there's a big difference behind the scenes: instead of immediately performing the operation, dtplyr just records what you're doing so when needed it can generate a single, efficient, data.table statement. You can preview the transformation (including the generated data.table code) by printing the result:
+
+```{r}
+mtcars2 %>% 
+  filter(wt < 5) %>% 
+  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+  group_by(cyl) %>% 
+  summarise(l100k = mean(l100k))
+```
+
+Generally, however, you should reserve this for exploration and debugging, and instead use `as.data.table()`, `as.data.frame()`, or `as_tibble()` to indicate that you're done writing the transformation and want to access the results:
```

Comment:
```suggestion
Generally, however, you should reserve printing for exploration and debugging, and instead use `as.data.table()`, `as.data.frame()`, or `as_tibble()` to indicate that you're done writing the transformation and want to access the results:
```

Reply:
Or some better word—just replace the "this."

Reply:
I went with "this preview"